src/Gonimo/Server/AuthHandlers/Internal.hs

```diff
@@ -0,0 +1,47 @@
+module Gonimo.Server.AuthHandlers.Internal where
+
+import           Control.Concurrent.STM   (STM, TVar, readTVar)
+import           Control.Lens             ((^.))
+import           Control.Monad.Freer      (Eff)
+import qualified Data.Map.Strict          as M
+import qualified Data.Set                 as S
+import           Gonimo.Server.Auth       (AuthServerConstraint, authorize,
+                                           authorizeAuthData, authorizeJust,
+                                           clientKey, isFamilyMember)
+import           Gonimo.Server.DbEntities (ClientId, FamilyId)
+import           Gonimo.Server.Effects    (atomically, getState)
+import           Gonimo.Server.State      (FamilyOnlineState, onlineMembers)
+
+
+
+authorizedPut :: AuthServerConstraint r
+              => (ClientId -> ClientId -> TVar FamilyOnlineState -> STM ())
+              ->  FamilyId -> ClientId -> ClientId -> Eff r ()
+authorizedPut f familyId fromId toId = do
+  authorizeAuthData (isFamilyMember familyId)
+  authorizeAuthData ((fromId ==) . clientKey)
+
+  familyOnlineData  <- authorizeJust (familyId `M.lookup`)
+                    =<< atomically . readTVar
+                    =<< getState
+
+  familyOnlineState <- atomically $ readTVar familyOnlineData
+  let fromto = S.fromList [fromId, toId]
+
+  authorize (fromto `S.isSubsetOf`) (familyOnlineState^.onlineMembers)
+
+  atomically $ f fromId toId familyOnlineData
+
```

Comment:
It looks ok in this case - because as far as I understand the code you only read TVars and never write. (Which makes sense in authcode, but in general be very carefull by splitting up code in multiple atomically statements. It is probably better to retrieve everything in one single atomically swoop and have the rest be pure code operating on the retrieved data - This should be a working approach, especially for auth code.

If you split things up in multiple calls to atomically you have to be very careful that you operate on a consistent view. -> General rule: Make pure as much as possible.